src/geos_types.jl

```diff
@@ -74,20 +74,34 @@ end
 
 mutable struct Polygon <: AbstractGeometry
     ptr::GEOSGeom
-
+    # create polygon using GEOSGeom pointer - only makes sense if pointer points to a polygon or a linear ring to start with. 
     function Polygon(ptr::GEOSGeom)
-        polygon = new(ptr)
+        id = LibGEOS.geomTypeId(ptr)
+        if id == 3
+            polygon = new(cloneGeom(ptr))
+        elseif id == 2
+            polygon = new(cloneGeom(createPolygon(ptr)))
```

Comment:
Is the cloneGeom needed here? Not sure if LibGEOS takes care of that down the line in createPolygon.

Reply:
@evetion I also thought that and didn't have it there initially because I assumed that `createPolygon` would take care of that through the `GEOSGeom_createPolygon_r` call. However, my tests were throwing a seg fault without the cloneGeom there. I can try it again, but my conclusion was that it is necessary. However, I did find that concerning because I also assumed that`GEOSGeom_createPolygon_r` would take care of it given the description within the [C API](https://libgeos.org/doxygen/geos__c_8h.html#a045b4db52c21a863b8f66c9e100317ad) that it creates "a newly allocated geometry." Thoughts on how to test this out?

Reply:
Looking into it, it does look like we should clone them for safety because the created polygon/geometry will take ownership of the input arguments (see e.g. https://trac.osgeo.org/geos/ticket/1111).

My comments are (embarrassingly) confusingly worded in https://github.com/JuliaGeo/LibGEOS.jl/blob/2c83c281b31c6592f8d4c78c2cc341b1b4bb1811/src/geos_functions.jl#L394 and https://github.com/JuliaGeo/LibGEOS.jl/blob/2c83c281b31c6592f8d4c78c2cc341b1b4bb1811/src/geos_functions.jl#L432-L434

Reply:
Okay, that makes sense @yeesian. Two more questions on this and then I will fix up these comments and submit it again first thing Monday morning.

(1) Do we think we should be using cloneGeom for the case where the polygon is made from coordinate vectors (lines 92-98)? I don't think we need to given that the pointers `interiors` and `exterior` aren't used outside of the constructor so it is okay that the new geometry takes ownership of them. Does this logic seem correct or do you think it is safer to make a copy?

(2) Do you think this might have larger impacts on other parts of the code? For example, what about `intersection` that calls `GEOSIntersection_r()`. Do you think we should be cloning the output? Could this explain issue #91?

Reply:
> (1) Do we think we should be using cloneGeom for the case where the polygon is made from coordinate vectors (lines 92-98)? I don't think we need to given that the pointers interiors and exterior aren't used outside of the constructor so it is okay that the new geometry takes ownership of them. Does this logic seem correct or do you think it is safer to make a copy?

Yeah, the logic seems sound: `interiors` and `exterior` are created within the scope of that function, so it's safe. I also doublechecked memory ownership of the vectors being used to construct the coordinate sequences and those are fine too (see [[1]](https://github.com/JuliaGeo/LibGEOS.jl/blob/2c83c281b31c6592f8d4c78c2cc341b1b4bb1811/src/geos_functions.jl#L46-L58), [[2]](https://github.com/JuliaGeo/LibGEOS.jl/blob/2c83c281b31c6592f8d4c78c2cc341b1b4bb1811/src/geos_functions.jl#L194-L208)).

> (2) Do you think this might have larger impacts on other parts of the code? For example, what about intersection that calls GEOSIntersection_r(). Do you think we should be cloning the output? Could this explain issue https://github.com/JuliaGeo/LibGEOS.jl/issues/91?

To my understanding, the output for `GEOSIntersection_r()` will be a new object and not require cloning. Unfortunately, I think it's likely that we'll have to do this analysis on a case-by-case for the other functions in the code (e.g. I think it might be argued that we should do it for [`getGeometries`](https://github.com/JuliaGeo/LibGEOS.jl/blob/2c83c281b31c6592f8d4c78c2cc341b1b4bb1811/src/geos_operations.jl#L236-L239) and [interior/exteriorRings](https://github.com/JuliaGeo/LibGEOS.jl/blob/2c83c281b31c6592f8d4c78c2cc341b1b4bb1811/src/geos_operations.jl#L244-L246)).

From quickly re-reading the code in this package, my hunch is that it's mainly a risk in this package when dealing with containers of geometries (so e.g. `multi*`, `geometrycollection`, and `polygon`).